### PR TITLE
Introduces subscription background color.

### LINKED ISSUE #130: add the ability to select a background/primary color
### Background
We will refer to the background color as the primary color. This is going to be used on the subscription card both on home and monthly page. This is a heavier task and will make changes all the way from the UI to the database so if you feel like its too large or you only want to do a part of it, make a sub-task or let me know and ill make one.

### TODO
- update cards background color on home and monthly page
- add a select list on upsert where you have  
    - only the following colors should be allowed (add a public readonly ienumerable of strings prop inside CssUtil): 
#theDefaultColor i think its #222
#281439
#34126d
#38354a
#22202c
#4b013b
#141c39
#34424c
#203d50
#13516d
#050505
#242424
#2e2e2e
#2d0903
    - It would be nice if the select list was actually grid styled so you don't have to scroll a long list of colors but it is not a requirement. default value should be should be the previously default value (think its #222 but you can go to bootswatch link in readme and inspect the fonts color using a chrome extension like eyedropper).
    - Each select item is basically a rectangle with a background of the color and with the entered subscription title, if its difficult for some reason to use the subscription title just hardcode an example text such as "lorem" or literally "example", "sample"
- On upsert move the description input field to the bottom of the page as showed in the mockups
- add the non nullable string field to the Subscription entity and when making the migration make sure to write hasdefault value inside the migration https://stackoverflow.com/questions/31025967/code-first-migration-how-to-set-default-value-for-new-property
- SubscriptionService when calling update or create the field should be mapped so that it is actually stored in the database

### Acceptance Criteria
- Existing users will have default colors of the previous colors
- When upserting a subscription we can select a primary color but it should not be required
- The subscription cards on home and monthly page has their background color set to the primary color

### PR DESCRIPTION
**Pull request by ashwin003**
Closes #130

### CODE REVIEW

File: subtrack.DAL/Migrations/SubtrackDbContextModelSnapshot.cs [hunk 1]
```diff
@@ -15,7 +15,7 @@ partial class SubtrackDbContextModelSnapshot : ModelSnapshot
         protected override void BuildModel(ModelBuilder modelBuilder)
         {
 #pragma warning disable 612, 618
-            modelBuilder.HasAnnotation("ProductVersion", "6.0.18");
+            modelBuilder.HasAnnotation("ProductVersion", "7.0.10");
```
Comment (chrisK00) [inline subtrack.DAL/Migrations/SubtrackDbContextModelSnapshot.cs]: I don't think this is supposed to be included

**Comment by chrisK00** [comment]
- The background color isnt displayed on the monthly/home page cards.
- Can we make the background color text a bit more inline/on the same row with the selector and reduce the width on the selector?
qemu-system-x86_64_1ldg62Bmf7
- Rename background color to Primary Color
- On upsert move the description input field to the bottom of the page as showed in the mockups. Billing cycle should be above primary color.
- I cannot create new subscriptions, the BackgroundColor prop is set to null when I havent selected a color

File: subtrack.DAL/Migrations/SubtrackDbContextModelSnapshot.cs [hunk 2]
```diff
@@ -66,7 +69,7 @@ protected override void BuildModel(ModelBuilder modelBuilder)
 
                     b.Property<string>("Name")
                         .IsRequired()
-                        .HasMaxLength(12)
+                        .HasMaxLength(50)
```
Comment (chrisK00) [inline subtrack.DAL/Migrations/SubtrackDbContextModelSnapshot.cs:67]: was this changed manually or is it just github messing around?

File: subtrack.MAUI/Utilities/CssUtil.cs [hunk 3]
```diff
@@ -2,6 +2,23 @@
 
 public static class CssUtil
 {
+    internal static IEnumerable<string> AvailableBackgroundColors = new[]
```
Comment (chrisK00) [inline subtrack.MAUI/Utilities/CssUtil.cs]:
```suggestion
    internal static readonly IEnumerable<string> AvailableBackgroundColors = new[]
```

**Review by chrisK00 — CHANGES_REQUESTED** [review]

File: subtrack.DAL/Migrations/SubtrackDbContextModelSnapshot.cs (hunk 2 above)
Comment (ashwin003) [reply subtrack.DAL/Migrations/SubtrackDbContextModelSnapshot.cs:67]: auto generated during migration

File: subtrack.DAL/Migrations/SubtrackDbContextModelSnapshot.cs (hunk 1 above)
Comment (ashwin003) [reply subtrack.DAL/Migrations/SubtrackDbContextModelSnapshot.cs]: also autogenerated by the migration

Comment (chrisK00) [reply subtrack.DAL/Migrations/SubtrackDbContextModelSnapshot.cs]: > also autogenerated by the migration

Yeah isnt it cuz you are running the app in .net 7?

Comment (ashwin003) [reply subtrack.DAL/Migrations/SubtrackDbContextModelSnapshot.cs]: probably, yes

**Comment by ashwin003** [comment]
> * The background color isnt displayed on the monthly/home page cards.
> * Can we make the background color text a bit more inline/on the same row with the selector and reduce the width on the selector?
>   qemu-system-x86_64_1ldg62Bmf7
> * Rename background color to Primary Color
> * On upsert move the description input field to the bottom of the page as showed in the mockups. Billing cycle should be above primary color.
> * I cannot create new subscriptions, the BackgroundColor prop is set to null when I havent selected a color

Should work now

File: subtrack.DAL/Migrations/SubtrackDbContextModelSnapshot.cs (hunk 1 above)
Comment (ashwin003) [reply subtrack.DAL/Migrations/SubtrackDbContextModelSnapshot.cs]: fixed it

File: subtrack.DAL/Entities/Subscription.cs [hunk 4]
```diff
@@ -26,6 +26,8 @@ public class Subscription : ICloneable
     [Range(1, int.MaxValue, ErrorMessage = "Interval has to be greater than 0")]
     public int BillingInterval { get; set; }
 
+    public string BackgroundColor { get; set; }
```
Comment (chrisK00) [inline subtrack.DAL/Entities/Subscription.cs]: Primary Color

Comment (chrisK00) [inline subtrack.DAL/Entities/Subscription.cs]: On the home page and monthly, display the color as the background color for the card - style background-color:@Sub.PrimaryColor

**Review by chrisK00 — CHANGES_REQUESTED** [review]

File: subtrack.MAUI/Pages/UpsertSubscription.razor [hunk 5]
```diff
@@ -24,28 +24,39 @@
...
+    <div style="display: flex;">
+        <label style="white-space: nowrap; padding-top: 5px; padding-right: 10px; font-size: 17px;">Last Payment </label>
+        <InputDate class="form-control" @bind-Value="@_subscription.LastPayment"></InputDate>
+    </div>
+    <div class="form-check form-switch" style="padding-top: 10px">
+        <label class="form-check-label" for="autopaid">Auto Paid </label>
+        <InputCheckbox class="form-check-input" id="autopaid" @bind-Value="@_subscription.IsAutoPaid"></InputCheckbox>
+    </div>
+    <div style="display: flex;">
+        <label style="white-space: nowrap; padding-top: 5px; padding-right: 10px; font-size: 17px;" for="backgroundColor">Primary Color </label>
```
Comment (chrisK00) [inline subtrack.MAUI/Pages/UpsertSubscription.razor]:
```suggestion
        <label style="white-space: nowrap; padding-right: 10px; font-size: 17px;" for="backgroundColor">Primary Color </label>
```

its still missaligned, due to the padding-top

File: subtrack.MAUI/Pages/UpsertSubscription.razor [hunk 6]
```diff
@@ -24,28 +24,39 @@
...
         <InputNumber class="form-control me-4" placeholder="Billing Interval" @bind-Value="@_subscription.BillingInterval" id="billingInterval" style=" flex: 2 1 6em"></InputNumber>
-         <InputSelect @bind-Value="@_subscription.BillingOccurrence" class="form-control flex-grow-1">
+        <InputSelect @bind-Value="@_subscription.BillingOccurrence" class="form-control flex-grow-1">
             <option value="@BillingOccurrence.Week">@GetBillingOccurrenceDisplayText("Week")</option>
             <option value="@BillingOccurrence.Month">@GetBillingOccurrenceDisplayText("Month")</option>
             <option value="@BillingOccurrence.Year">@GetBillingOccurrenceDisplayText("Year")</option>
         </InputSelect>
     </div>
+    <br />
+    <div style="display: flex;">
```
Comment (chrisK00) [inline subtrack.MAUI/Pages/UpsertSubscription.razor]:
```suggestion
    <div style="display: flex;" class="mt-2">
```

File: subtrack.MAUI/Pages/UpsertSubscription.razor [hunk 7]
```diff
@@ -24,28 +24,39 @@
...
+        <label style="white-space: nowrap; padding-top: 5px; padding-right: 10px; font-size: 17px;" for="backgroundColor">Primary Color </label>
+        <input class="form-control" type="color" id="backgroundColor" list="backgroundPresets" value="@_subscription.BackgroundColor" @onchange="@UpdateBackgroundColor" />
+        <datalist id="backgroundPresets">
+            @foreach (var backgroundColor in backgroundColorOptions)
+            {
+                <option value="@backgroundColor.ToLower()">@backgroundColor.ToLower()</option>
+            }
+        </datalist>
+    </div>
+    <div class="form-group" style="padding-top: 10px;">
```
Comment (chrisK00) [inline subtrack.MAUI/Pages/UpsertSubscription.razor]:
```suggestion
    <div class="form-group mt-2" style="padding-top: 10px;">
```

File: subtrack.MAUI/Pages/UpsertSubscription.razor [hunk 8]
```diff
@@ -24,28 +24,39 @@
...
+        <label style="white-space: nowrap; padding-top: 5px; padding-right: 10px; font-size: 17px;">Last Payment </label>
+        <InputDate class="form-control" @bind-Value="@_subscription.LastPayment"></InputDate>
+    </div>
+    <div class="form-check form-switch" style="padding-top: 10px">
+        <label class="form-check-label" for="autopaid">Auto Paid </label>
+        <InputCheckbox class="form-check-input" id="autopaid" @bind-Value="@_subscription.IsAutoPaid"></InputCheckbox>
+    </div>
+    <div style="display: flex;">
+        <label style="white-space: nowrap; padding-top: 5px; padding-right: 10px; font-size: 17px;" for="backgroundColor">Primary Color </label>
+        <input class="form-control" type="color" id="backgroundColor" list="backgroundPresets" value="@_subscription.BackgroundColor" @onchange="@UpdateBackgroundColor" />
```
Comment (chrisK00) [inline subtrack.MAUI/Pages/UpsertSubscription.razor]:
```suggestion
        <input class="form-control" type="color" style="max-width: 100px;min-height:25px" id="backgroundColor" list="backgroundPresets" value="@_subscription.BackgroundColor" @onchange="@UpdateBackgroundColor" />
```

**Review by chrisK00 — CHANGES_REQUESTED** [review]

**Review by chrisK00 — APPROVED** [review]

File: subtrack.DAL/Entities/Subscription.cs (hunk 4 above)
Comment (ashwin003) [reply subtrack.DAL/Entities/Subscription.cs]: Done

Comment (ashwin003) [reply subtrack.DAL/Entities/Subscription.cs]: Done. With a separate migration.

**Pull request merged**